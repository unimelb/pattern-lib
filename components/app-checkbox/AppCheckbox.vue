<template>
  <label
    :data-name="name"
    class="app-checkbox"
    @click.prevent="onClick">
    <input
      :aria-checked="isInputChecked"
      :aria-label="ariaLabel"
      :checked="isInputChecked"
      class="app-checkbox__input"
      data-testid="app-checkbox-input"
      type="checkbox">

    <span class="app-checkbox__icon">
      <span
        :class="classIconChecked"
        data-testid="app-checkbox-icon-checked" />
      <span
        :class="classIconIndeterminate"
        data-testid="app-checkbox-icon-indeterminate" />
    </span>
    <span
      v-if="!!label"
      class="app-checkbox__label"
      data-testid="app-checkbox-label">{{ label }}</span>
  </label>
</template>

<script>
export default {
  name: 'AppCheckbox',
  model: {
    event: 'change',
  },
  props: {
    name: {
      type: String,
      required: true,
    },
    label: {
      type: String,
      default: '',
    },
    isChecked: {
      type: Boolean,
      default: false,
    },
    isIndeterminate: {
      type: Boolean,
      default: false,
    },
    ariaLabel: {
      type: String,
      default: '',
    },
  },
  computed: {
    isInputChecked() {
      return !this.isIndeterminate && this.isChecked;
    },
    classIconChecked() {
      return [
        'app-checkbox__icon-checked',
        {
          'app-checkbox__icon-checked--is-checked': this.isInputChecked,
        },
      ];
    },
    classIconIndeterminate() {
      return [
        'app-checkbox__icon-indeterminate',
        {
          'app-checkbox__icon-indeterminate--is-indeterminate': this.isIndeterminate,
        },
      ];
    },
  },
  methods: {
    onClick() {
      const { name, isChecked, isIndeterminate } = this;

      // Current values emited
      this.$emit('change', {
        name,
        isChecked,
        isIndeterminate,
      });
    },
  },
};
</script>
