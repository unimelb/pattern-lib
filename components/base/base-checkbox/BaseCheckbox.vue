<template>
  <label
    role="checkbox"
    :data-name="name"
    :class="classRoot"
    :tabindex="tabIndex"
    :aria-label="ariaLabel"
    :aria-checked="isInputChecked + ''"
    @focus="onRootFocus"
    @blur="onRootBlur"
    @click.prevent="toggleHandler"
    @keypress.space.prevent="toggleHandler">
    <input
      :checked="isInputChecked"
      class="base-checkbox__input"
      data-testid="base-checkbox-input"
      type="checkbox">

    <span :class="classIcon">
      <span
        :class="classIconChecked"
        data-testid="base-checkbox-icon-checked" />
      <span
        :class="classIconIndeterminate"
        data-testid="base-checkbox-icon-indeterminate" />
    </span>
    <span
      v-if="!!label"
      class="base-checkbox__label"
      data-testid="base-checkbox-label">{{ label }}</span>
  </label>
</template>

<script>
const displayProp = {
  inline: 'inline',
  block: 'block',
};

export default {
  name: 'BaseCheckbox',
  model: {
    event: 'change',
  },
  props: {
    name: {
      type: String,
      required: true,
    },
    label: {
      type: String,
      default: '',
    },
    isChecked: {
      type: Boolean,
      default: false,
    },
    isIndeterminate: {
      type: Boolean,
      default: false,
    },
    ariaLabel: {
      type: String,
      default: undefined,
    },
    display: {
      type: String,
      default: displayProp.inline,
      validator(value) {
        return Object.values(displayProp).includes(value);
      },
    },
    tabIndex: {
      type: Number,
      default: 0,
    },
  },
  data() {
    return {
      isFocused: false,
    };
  },
  computed: {
    isInputChecked() {
      return !this.isIndeterminate && this.isChecked;
    },
    classRoot() {
      return [
        'base-checkbox',
        `base-checkbox--${this.display}`,
      ];
    },
    classIcon() {
      return [
        'base-checkbox__icon',
        {
          'base-checkbox__icon--is-focused': this.isFocused,
        },
      ];
    },
    classIconChecked() {
      return [
        'base-checkbox__icon-checked',
        {
          'base-checkbox__icon-checked--is-checked': this.isInputChecked,
        },
      ];
    },
    classIconIndeterminate() {
      return [
        'base-checkbox__icon-indeterminate',
        {
          'base-checkbox__icon-indeterminate--is-indeterminate': this.isIndeterminate,
        },
      ];
    },
  },
  methods: {
    toggleHandler() {
      const { name, isChecked, isIndeterminate } = this;

      // Current values emited
      this.$emit('change', {
        name,
        isChecked,
        isIndeterminate,
      });
    },
    onRootFocus() {
      this.isFocused = true;
    },
    onRootBlur() {
      this.isFocused = false;
    },
  },
};
</script>
