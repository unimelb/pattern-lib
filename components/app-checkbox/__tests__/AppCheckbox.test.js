import { mount } from 'vue-test-utils';
import AppCheckbox from '../AppCheckbox.vue';
import BaseCheckbox from '../../base/base-checkbox/BaseCheckbox';

describe('AppCheckbox', () => {
  describe('single modelValue', () => {
    describe('should render BaseCheckbox: ', () => {
      const cases = [
        [
          'all props',
          {
            name: 'app-checkbox',
            modelValue: true,
            label: 'App checkbox',
            ariaLabel: 'ariaLabel',
          },
          {
            name: 'app-checkbox',
            label: 'App checkbox',
            isChecked: true,
            isIndeterminate: false,
            ariaLabel: 'ariaLabel',
          },
        ],
        [
          'only required props',
          {
            modelValue: true,
          },
          {
            name: '',
            label: '',
            isChecked: true,
            isIndeterminate: false,
            ariaLabel: '',
          },
        ],
      ];

      it.each(cases)('%s', (description, propsData, expectedBaseProps) => {
        const wrapper = mount(AppCheckbox, {
          propsData,
        });

        const baseCheckbox = wrapper.find(BaseCheckbox);

        expect(baseCheckbox.props()).toEqual(expectedBaseProps);
      });
    });

    describe('should emit change event with a new value', () => {
      const cases = [
        [true],
        [false],
      ];

      it.each(cases)('checkbox value: %s', (modelValue) => {
        const propsData = {
          name: 'app-checkbox',
          modelValue,
          label: 'App checkbox',
          ariaLabel: 'ariaLabel',
        };

        const wrapper = mount(AppCheckbox, {
          propsData,
        });

        wrapper.find(BaseCheckbox).vm.$emit('change', { isChecked: modelValue });

        const emmitedChange = wrapper.emitted('change');
        expect(emmitedChange.length).toEqual(1);
        expect(emmitedChange[0].length).toEqual(1);
        expect(emmitedChange[0][0]).toEqual(!modelValue);
      });
    });
  });

  describe('modelValue - array of checked checkbox names', () => {
    it('should throw an error if the name prop is not passed', () => {
      let errorThrown = false;

      try {
        mount(AppCheckbox, {
          propsData: {
            modelValue: [],
          },
        });
      } catch (e) {
        errorThrown = true;
      }

      expect(errorThrown).toEqual(true);
    });

    describe('checkbox state', () => {
      const cases = [
        ['should have checked state', ['checkbox1', 'appCheckbox'], true],
        ['should have unchecked state', ['checkbox1'], false],
      ];

      it.each(cases)('%s', (description, modelValue, expectedCheckedState) => {
        const wrapper = mount(AppCheckbox, {
          propsData: {
            name: 'appCheckbox',
            modelValue,
          },
        });

        expect(wrapper.find(BaseCheckbox).props().isChecked).toEqual(expectedCheckedState);
      });
    });

    describe('should emit change event with a new value', () => {
      const cases = [
        [
          'should add the checkbox name to a modelValue',
          ['checkbox1'],
          ['checkbox1', 'appCheckbox'],
        ],
        [
          'should remove the checkbox name to a modelValue',
          ['checkbox1', 'appCheckbox'],
          ['checkbox1'],
        ],
      ];

      it.each(cases)('%s', (description, modelValue, expectedChangedValue) => {
        const wrapper = mount(AppCheckbox, {
          propsData: {
            name: 'appCheckbox',
            modelValue,
          },
        });

        wrapper.find(BaseCheckbox).vm.$emit('change', { isChecked: modelValue.includes('appCheckbox') });

        const emmitedChange = wrapper.emitted('change');
        expect(emmitedChange.length).toEqual(1);
        expect(emmitedChange[0][0]).toEqual(expectedChangedValue);
      });
    });
  });
});
