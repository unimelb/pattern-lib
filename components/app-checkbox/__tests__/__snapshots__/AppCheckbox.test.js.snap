// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppCheckbox should match snapshot 1`] = `
<label
  class="app-checkbox"
  data-name="checkboxName"
>
  <input
    aria-label="aria-label"
    class="app-checkbox__input"
    data-testid="app-checkbox-input"
    type="checkbox"
  />
   
  <span
    class="app-checkbox__icon"
  >
    <span
      class="app-checkbox__icon-checked"
      data-testid="app-checkbox-icon-checked"
    />
     
    <span
      class="app-checkbox__icon-indeterminate"
      data-testid="app-checkbox-icon-indeterminate"
    />
  </span>
   
  <span
    class="app-checkbox__label"
    data-testid="app-checkbox-label"
  >
    checkbox label
  </span>
</label>
`;
